### ISSUE TITLE
[heal] Missing part file in one disk is not detected by Heal

### ISSUE THREAD

**Issue by vadmeste** [minio/minio#3961]
## Expected Behavior
Heal should detect when xl.json is not found in one disk, because if two other disks fail (/ 4 disks), we won't be able able to recover anymore.

## Current Behavior
Heal doesnt detect missing xl.json

## Possible Solution
Reconstruct as we do with a replaced disk

## Steps to Reproduce (for bugs)

1. `minio server /tmp/xl{1..4}`
2. `mc mb myminio/testbucket ; mc cp 128mb.file myminio/testbucket`
3. `rm /tmp/xl1/testbucket/128mb.file/xl.json`
4. `mc admin heal myminio/testbucket/128mb.file`
Nothing

5. ` rm -rf /tmp/xl{2..4}`
6. `mc cat myminio/testbucket/128mb.file` => error

## Context
Testing heal

## Your Environment
All

**Comment by vadmeste** [comment]
I changed the title of this to indicate that the same goes for a part file, if it is missing, it is not detected by Heal.

**Comment by krisis** [comment]
@vadmeste The issue is not reproducible with minio version,
```
Version: 2017-03-24T04:04:21Z
Release-Tag: DEVELOPMENT.2017-03-24T04-04-21Z
Commit-ID: 8c0ce2fee9ec7b96533235800d6382beadd98505

```
```
$ mc mb myminio/testbucket ; mc cp 128mb.file myminio/testbucket 
mc: <ERROR> Unable to make bucket `myminio/testbucket`. Your previous request to create the named bucket succeeded and you already own it.
128mb.file:      128.00 MB / 128.00 MB ┃▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓┃ 100.00% 93.81 MB/s 1s

$ rm /tmp/xl1/testbucket/128mb.file/xl.json

$ mc admin heal myminio/testbucket/128mb.file
Object testbucket/128mb.file is healed.

```

**Comment by harshavardhana** [comment]
> minio server /tmp/xl{1..4}
> 
> mc mb myminio/testbucket ; mc cp 128mb.file myminio/testbucket
> 
> rm /tmp/xl1/testbucket/128mb.file/xl.json
> 
> mc admin heal myminio/testbucket/128mb.file
> Nothing
> 
> rm -rf /tmp/xl{2..4}

You have deleted 3 copies out of 4 - we need minimum of N/2 to read and re-construct.

**Comment by krisis** [comment]
>   You have deleted 3 copies out of 4 - we need minimum of N/2 to read and re-construct.
@harshavardhana IIUC, @vadmeste is demonstrating that if `mc admin heal` fails to detect the need to heal at an earlier stage, we could lose the data permanently. You are right, if the user/admin ever loses 3 out of 4 disks, we can't recover the data using `mc admin heal`.

**Comment by harshavardhana** [comment]
> @harshavardhana IIUC, @vadmeste is demonstrating that if mc admin heal fails to detect the need to heal at an earlier stage, we could lose the data permanently. You are right, if the user/admin ever loses 3 out of 4 disks, we can't recover the data using mc admin heal.

We can report that it is not possible to heal atleast.. But when i read the `Possible Solution` section. 

> Reconstruct as we do with a replaced disk

^^ this doesn't make sense to me.

**Comment by vadmeste** [comment]
> You have deleted 3 copies out of 4 - we need minimum of N/2 to read and re-construct.

@harshavardhana, Yes my bad, I meant deleting 2 out of 4.

> The issue is not reproducible with minio version,

@krisis, well, I wasn't able to reproduce with xl.json, but the problem is there with part files. If one part file is removed, it won't be detected by Heal.

> Reconstruct as we do with a replaced disk
> ^^ this doesn't make sense to me.

@harshavardhana: as @krisis mentioned, we need to detect that anomaly at early stage. If we have total 4 disks, and one object has a missing part file in one of the disks, losing two other disks will make that object unrecoverable. My suggested solution is to simply use the same healing process that we use with fresh disks, in other words, we can remove existing xl.json + all part files and reconstruct new ones.

**Comment by harshavardhana** [comment]
Is this fixed with #3967 ?

**Comment by krisis** [comment]
>Is this fixed with #3967 ?

Yes, it is fixed.

**Comment by harshavardhana** [comment]
> Yes, it is fixed.

Closing as fixed by #3967
